File: app/src/main/java/com/example/benedict/ConnectionApp/MainActivity.java
package com.example.benedict.ConnectionApp;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity implements View.OnTouchListener, MediaPlayer.OnCompletionListener {

    private TextView txtDuration;
    private ImageView imgSignal,imgMicrophone, imgPlay, imgStop, imgRefresh;
    private ManifestPermission manifestPermission;

    private Rect rect;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        init();
        manifestPermission = ManifestPermission.newInstance(this);
        Log.d("getMicrophoneImage",String.valueOf(getMicrophoneImage() == isMicrphoneNormal()));
        //TODO: imgSignal - should be equal to the input decibel source of sound in the microphone
        //TODO: imgRefresh - when clicked delete the recorded audio MediaRecorderSample.3gp
        //TODO: showAppPermissionSettings() - call this method when the permissions is not granted and checked the never ask again
    }

    @SuppressLint("ClickableViewAccessibility")
    private void init() {
        txtDuration = (TextView) findViewById(R.id.text_view_duration);
        imgSignal = (ImageView) findViewById(R.id.image_view_signal);
        imgRefresh = (ImageView) findViewById(R.id.image_view_refresh);
        imgPlay = (ImageView) findViewById(R.id.image_view_play);
        imgStop = (ImageView) findViewById(R.id.image_view_stop);
        imgMicrophone = (ImageView) findViewById(R.id.image_view_microphone);

        imgMicrophone.setOnTouchListener(this);
        imgRefresh.setOnTouchListener(this);
        imgPlay.setOnTouchListener(this);
        imgStop.setOnTouchListener(this);
    }

    @Override
    public boolean onTouch(View view, MotionEvent event) {
        float x = event.getX();
        float y = event.getY();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                // Write your code to perform an action on down
                rect = new Rect(view.getLeft(), view.getTop(), view.getRight(), view.getBottom());
                if (view.getId() == imgMicrophone.getId()) {
                    Log.d("onTouch","ACTION_DOWN imgMicrophone");
                    startRecord();
                } else if (view.getId() == imgRefresh.getId()) {
                    Log.d("onTouch","ACTION_DOWN imgRefresh");
                }
                else if (view.getId() == imgPlay.getId()) {
                    Log.d("onTouch","ACTION_DOWN imgPlay");
                }
                else if (view.getId() == imgStop.getId()) {
                    Log.d("onTouch","ACTION_DOWN imgStop");
                }
                return true;
            case MotionEvent.ACTION_MOVE:
                // Write your code to perform an action on continuous touch move
                if (view.getId() == imgMicrophone.getId()) {
                    Log.d("onTouch","ACTION_MOVE imgMicrophone");
                    startRecord();
                }
                return true;
            case MotionEvent.ACTION_UP:
                // Write your code to perform an action on touch up
                Log.d("onTouch","ACTION_UP");
                stopRecord();
                if (rect.contains(view.getLeft() + (int) x, view.getTop() + (int) y)) {
                    // User moved inside bounds
                    if (view.getId() == imgMicrophone.getId()) {
                        Log.d("onTouch","ACTION_UP imgMicrophone");
                        stopRecord();
                    } else if (view.getId() == imgRefresh.getId()) {
                        Log.d("onTouch","ACTION_UP imgRefresh");
                        resetRecord();
                    }
                    else if (view.getId() == imgPlay.getId()) {
                        Log.d("onTouch","ACTION_UP imgPlay");
                        playAudio();
                    }
                    else if (view.getId() == imgStop.getId()) {
                        Log.d("onTouch","ACTION_UP imgStop");
                        stopAudio();
                    }
                    return true;
                }
                return false;
        }
        return false;
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        stopAudio();
    }

    @Override
    protected void onResume() {
        super.onResume();
        manifestPermission.check();
    }
    //region Presenter
    private void startRecord() {
        if (getMicrophoneImage() == isMicrphoneNormal()) {
            AudioRecorder.start();
            startRecordTimerAsyncTask();
        }
    }

    private void stopRecord() {
        if (getMicrophoneImage() == isMicrphonePressed()) {
            AudioRecorder.stop();
            AudioRecorder.release();
            resetRecordView();
        }
    }

    private void playAudio() {
        AudioPlayer.start(this);
        playAudioVisibility();
        startPlayTimerAsyncTask();
    }

    private void stopAudio() {
        AudioPlayer.stop();
        AudioPlayer.release();
        stopRecordView();
    }

    private void resetRecord() {
        stopAudio();
        setRecordVisibility();
        resetRecordView();
    }

    private void startRecordTimerAsyncTask() {
        RecordTimerAsyncTask.newInstance(this);
        RecordTimerAsyncTask.execute();
    }

    private void startPlayTimerAsyncTask() {
        PlayTimerAsyncTask.newInstance(this);
        PlayTimerAsyncTask.execute();
    }
    //endregion
    //region View Animation
    private void stopRecordView() {
        setTimerDurationColour(false);
        stopAudioVisibility();
    }

    private void resetRecordView() {
        setTimerDurationText("0");
        setTimerDurationColour(false);
        setMicrophoneImage(false);
    }

    public void enableRecordView() {
        setTimerDurationText("00:00");
        imgRefresh.setVisibility(View.INVISIBLE);
        imgMicrophone.setVisibility(View.VISIBLE);
        imgPlay.setVisibility(View.INVISIBLE);
        imgStop.setVisibility(View.INVISIBLE);
    }

    public void disableRecordView() {
        setTimerDurationText("00:00 - No Microphone");
        imgRefresh.setVisibility(View.INVISIBLE);
        imgMicrophone.setVisibility(View.INVISIBLE);
        imgPlay.setVisibility(View.INVISIBLE);
        imgStop.setVisibility(View.INVISIBLE);
    }

    private void setRecordVisibility() {
        imgRefresh.setVisibility(View.INVISIBLE);
        imgMicrophone.setVisibility(View.VISIBLE);
        imgPlay.setVisibility(View.INVISIBLE);
        imgStop.setVisibility(View.INVISIBLE);
    }

    public void setPlayVisibility() {
        imgRefresh.setVisibility(View.VISIBLE);
        imgMicrophone.setVisibility(View.INVISIBLE);
        stopAudioVisibility();
    }

    private void stopAudioVisibility() {
        imgPlay.setVisibility(View.VISIBLE);
        imgStop.setVisibility(View.INVISIBLE);
    }

    private void playAudioVisibility() {
        imgPlay.setVisibility(View.INVISIBLE);
        imgStop.setVisibility(View.VISIBLE);
    }

    public Boolean isPlayVisible() {
        return imgPlay.getVisibility() == View.VISIBLE;
    }

    public Boolean isStopVisible() {
        return imgStop.getVisibility() == View.VISIBLE;
    }

    public void setMicrophoneImage(Boolean isRecording) {
        if (isRecording) {
            imgMicrophone.setImageResource(R.drawable.ic_microphone_pressed);
        } else {
            imgMicrophone.setImageResource(R.drawable.ic_microphone_normal);
        }
    }

    public Drawable.ConstantState getMicrophoneImage() {
        return imgMicrophone.getDrawable().getConstantState();
    }

    private Drawable.ConstantState isMicrphoneNormal() {
        return getResources().getDrawable(
                R.drawable.ic_microphone_normal
        ).getConstantState();
    }

    public Drawable.ConstantState isMicrphonePressed() {
        return getResources().getDrawable(
                R.drawable.ic_microphone_pressed
        ).getConstantState();
    }

    public void setTimerDurationText(String data) {
        txtDuration.setText(data);
    }

    public void setTimerDurationColour(Boolean isRecording) {
        if (isRecording) {
            txtDuration.setTextColor(ContextCompat.getColor(getBaseContext(),R.color.green));
        } else {
            txtDuration.setTextColor(ContextCompat.getColor(this,R.color.black));
        }
    }
    //endregion
    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (ManifestPermission.RECORD_AUDIO == requestCode) {
            Log.d("PermissionsResult", "requestCode" + requestCode);
            Log.d("PermissionsResult", "permissions" + permissions);
            Log.d("PermissionsResult", "grantResults" + grantResults);
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this,"Permission granted successfully", Toast.LENGTH_SHORT).show();
                manifestPermission.setRequestGranted();
                enableRecordView();
            }  else {
                disableRecordView();
                //showAppPermissionSettings();
            }
        }
    }

    public void showAppPermissionSettings() {
        Toast.makeText(this,"Hardware Permissions Disabled", Toast.LENGTH_LONG).show();
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                Uri.fromParts("package", getPackageName(), null));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        AudioRecorder.release();
        AudioPlayer.release();
    }
}

File: app/src/main/java/com/example/benedict/permission/MainActivity.kt
package com.example.benedict.permission

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), View.OnClickListener {

    companion object {
        private val TAG : String = MainActivity::class.java.getSimpleName()
    }

    override fun onCreate(savedInstanceState : Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //region Check Permission On Click Listener
        allCheckPermissions.setOnClickListener(this@MainActivity)
        telephonyCheckPermissions.setOnClickListener(this@MainActivity)
        microphoneCheckPermissions.setOnClickListener(this@MainActivity)
        cameraCheckPermissions.setOnClickListener(this@MainActivity)
        videoCallCheckPermissions.setOnClickListener(this@MainActivity)
        galleryCheckPermissions.setOnClickListener(this@MainActivity)
        contactCheckPermissions.setOnClickListener(this@MainActivity)
        //endregion
        //region Request Permission On Click Listener
        allRequestPermissions.setOnClickListener(this@MainActivity)
        telephonyRequestPermissions.setOnClickListener(this@MainActivity)
        microphoneRequestPermissions.setOnClickListener(this@MainActivity)
        cameraRequestPermissions.setOnClickListener(this@MainActivity)
        videoCallRequestPermissions.setOnClickListener(this@MainActivity)
        galleryRequestPermissions.setOnClickListener(this@MainActivity)
        contactRequestPermissions.setOnClickListener(this@MainActivity)
        //endregion
        showRationalDialog.setOnClickListener(this@MainActivity)
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onClick(view : View) {
        when(view) {
            allCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.allPermissions,
                    isGranted = { setTextData("All Permissions Granted!") },
                    isDenied = { setTextData("All Permissions Denied!") }
                )
            }
            telephonyCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.telephonyPermissions,
                    isGranted = { setTextData("Telephony Permissions Granted!") },
                    isDenied = { setTextData("Telephony Permissions Denied!") }
                )
            }
            microphoneCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.microphonePermission,
                    isGranted = { setTextData("Microphone Permissions Granted!") },
                    isDenied = { setTextData("Microphone Permissions Denied!") }
                )
            }
            cameraCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.cameraPermission,
                    isGranted = { setTextData("Camera Permissions Granted!") },
                    isDenied = { setTextData("Camera Permissions Denied!") }
                )
            }
            videoCallCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.videoCallPermission,
                    isGranted = { setTextData("Video Call Permissions Granted!") },
                    isDenied = { setTextData("Video Call Permissions Denied!") }
                )
            }
            galleryCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.galleryPermissions,
                    isGranted = { setTextData("Gallery Permissions Granted!") },
                    isDenied = { setTextData("Gallery Permissions Denied!") }
                )
            }
            contactCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.contactPermission,
                    isGranted = { setTextData("Contact Permissions Granted!") },
                    isDenied = { setTextData("Contact Permissions Denied!") }
                )
            }
            allRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.allPermissions,
                    ManifestPermission.ALL_PERMISSION_CODE
                )
            }
            telephonyRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.telephonyPermissions,
                    ManifestPermission.TELEPHONY_PERMISSION_CODE
                )
            }
            microphoneRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.microphonePermission,
                    ManifestPermission.MICROPHONE_PERMISSION_CODE
                )
            }
            cameraRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.cameraPermission,
                    ManifestPermission.CAMERA_PERMISSION_CODE
                )
            }
            videoCallRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.videoCallPermission,
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE
                )
            }
            galleryRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.galleryPermissions,
                    ManifestPermission.GALLERY_PERMISSION_CODE
                )
            }
            contactRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.contactPermission,
                    ManifestPermission.CONTACT_PERMISSION_CODE
                )
            }
            showRationalDialog -> {
                ManifestPermission.showRationalDialog(this@MainActivity,"Go to App Permission Settings?")
            }
        }
    }

    private fun setTextData(text : String) {
        txtData.setText(text)
    }

    override fun onRequestPermissionsResult(requestCode : Int, permissions : Array<String>, grantResults : IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        Log.d("PermissionsResult", "requestCode $requestCode")
        Log.d("PermissionsResult","permissions " + permissions.contentToString())
        Log.d("PermissionsResult","grantResults " + grantResults.contentToString())

        permissions.map { permission ->
            ManifestPermission.checkNeverAskAgain(this@MainActivity,permission,
                isNeverAskAgain = {
                    Log.d("PermissionsResult", "checkNeverAskAgain isNeverAskAgain single $permission")
                },
                isNotNeverAskAgain = {
                    Log.d("PermissionsResult", "checkNeverAskAgain isNotNeverAskAgain single $permission")
                }
            )

            ManifestPermission.checkPermissionsResult(this@MainActivity,permission,
                isNeverAskAgain = {
                    Log.d("PermissionsResult", "checkPermissionsResult isNeverAskAgain single $permission")
                }, isDenied = {
                    Log.d("PermissionsResult","checkPermissionsResult isDenied single $permission")
                }, isGranted = {
                    Log.d("PermissionsResult","checkPermissionsResult isGranted single $permission")
                }
            )
        }

        ManifestPermission.checkNeverAskAgain(this@MainActivity, permissions,
            isNeverAskAgain = {
                Log.d("PermissionsResult", "checkNeverAskAgain isNeverAskAgain")
            },
            isNotNeverAskAgain = {
                Log.d("PermissionsResult", "checkNeverAskAgain isNotNeverAskAgain")
            }
        )

        ManifestPermission.checkPermissionsResult(this@MainActivity, permissions, grantResults,
            isNeverAskAgain =  {
                Log.d("PermissionsResult", "checkPermissionsResult isNeverAskAgain")
                when(requestCode) {
                    ManifestPermission.ALL_PERMISSION_CODE -> { setTextData("All Permissions is Never Ask Again!") }
                    ManifestPermission.TELEPHONY_PERMISSION_CODE -> { setTextData("Telephony Permissions is Never Ask Again!") }
                    ManifestPermission.MICROPHONE_PERMISSION_CODE -> { setTextData("Microphone Permissions is Never Ask Again!") }
                    ManifestPermission.CAMERA_PERMISSION_CODE -> { setTextData("Camera Permissions is Never Ask Again!") }
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE -> { setTextData("Video Call Permissions is Never Ask Again!") }
                    ManifestPermission.GALLERY_PERMISSION_CODE -> { setTextData("Gallery Permissions is Never Ask Again!") }
                    ManifestPermission.CONTACT_PERMISSION_CODE -> { setTextData("Contact Permissions is Never Ask Again!") }
                }
            }, isDenied = {
                Log.d("PermissionsResult","checkPermissionsResult isDenied")
                when(requestCode) {
                    ManifestPermission.ALL_PERMISSION_CODE -> { setTextData("All Permissions Denied!") }
                    ManifestPermission.TELEPHONY_PERMISSION_CODE -> { setTextData("Telephony Permissions Denied!") }
                    ManifestPermission.MICROPHONE_PERMISSION_CODE -> { setTextData("Microphone Permissions Denied!") }
                    ManifestPermission.CAMERA_PERMISSION_CODE -> { setTextData("Camera Permissions Denied!") }
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE -> { setTextData("Video Call Permissions Denied!") }
                    ManifestPermission.GALLERY_PERMISSION_CODE -> { setTextData("Gallery Permissions Denied!") }
                    ManifestPermission.CONTACT_PERMISSION_CODE -> { setTextData("Contact Permissions Denied!") }
                }
            }, isGranted = {
                Log.d("PermissionsResult","checkPermissionsResult isGranted")
                when(requestCode) {
                    ManifestPermission.ALL_PERMISSION_CODE -> { setTextData("All Permissions Granted!") }
                    ManifestPermission.TELEPHONY_PERMISSION_CODE -> { setTextData("Telephony Permissions Granted!") }
                    ManifestPermission.MICROPHONE_PERMISSION_CODE -> { setTextData("Microphone Permissions Granted!") }
                    ManifestPermission.CAMERA_PERMISSION_CODE -> { setTextData("Camera Permissions Granted!") }
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE -> { setTextData("Video Call Permissions Granted!") }
                    ManifestPermission.GALLERY_PERMISSION_CODE -> { setTextData("Gallery Permissions Granted!") }
                    ManifestPermission.CONTACT_PERMISSION_CODE -> { setTextData("Contact Permissions Granted!") }
                }
            }
        )
        ManifestPermission.checkPermissionsResult(this@MainActivity, ManifestPermission.ALL_PERMISSION_CODE, requestCode, permissions, grantResults,
            isNeverAskAgain = {
                Log.d("PermissionsResult","checkPermissionsResult with Selected Request Code, isNeverAskAgain")
            }, isDenied = {
                Log.d("PermissionsResult","checkPermissionsResult with Selected Request Code, isDenied")
            }, isGranted = {
                Log.d("PermissionsResult","checkPermissionsResult with Selected Request Code, isGranted")
            })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        Log.d(TAG,"onActivityResult($requestCode,$resultCode,$data)")
        if (requestCode == ManifestPermission.SETTINGS_PERMISSION_CODE)
            Toast.makeText(this@MainActivity,"PERMISSION_SETTINGS_CODE",Toast.LENGTH_SHORT).show()
    }

    override fun onDestroy() {
        Log.d(TAG,"onDestroy()")
        super.onDestroy()
    }
}